controller/array_action/ds8k_rest_client.py

```diff
@@ -0,0 +1,253 @@
+from pyds8k.exceptions import NotFound
+from controller.common.csi_logger import get_stdout_logger
+
+logger = get_stdout_logger()
+
+
+def normalize_response(func):
+    def inner(self, *args, **kwargs):
+        res_obj = func(self, *args, **kwargs)
+        if not isinstance(res_obj, list):
+            res_obj = [res_obj, ]
+        return list(map(_normalize_item, res_obj))
+    return inner
+
+
+def _normalize_item(res):
+    try:
+        return res.representation
+    except AttributeError:
+        return res
+
+
+def _int_lunid_to_hex(lunid):
+    return '{0:0{1}x}'.format(int(lunid), 2)
+
+
+def int_to_scsilun(lun):
+    """
+    There are two style lun number, one's decimal value is <256 and the other
+    is full as 16 hex digit. According to T10 SAN, when decimal value is more
+    than 256 and it is converted to the full 16 hex digit, it should be
+    swapped and converted into hex.
+    For example, zlinux lun number stored in SC DB is 1075331113. It should
+    be hex to '40184029' and swapped to '40294018'. When the lun number is 12
+    and it is <256, it should be hex to '0c' directly.
+
+    https://github.com/kubernetes/kubernetes/issues/45024
+    """
+    pretreated_lun = int(lun)
+    if pretreated_lun < 256:
+        return _int_lunid_to_hex(pretreated_lun)
+    else:
+        return '{0:x}'.format(
+            (pretreated_lun >> 16 & 0xFFFF) | (pretreated_lun & 0xFFFF) << 16
+        )
+
+
+def scsilun_to_int(lun):
+    """
+    There are two style lun number, one's decimal value is <256 and the other
+    is full as 16 hex digit. According to T10 SAN, the full 16 hex digit
+    should be swapped and converted into decimal.
+    For example, SC got zlinux lun number from DS8K API, '40294018'. And it
+    should be swapped to '40184029' and converted into decimal, 1075331113.
+    When the lun number is '0c' and its decimal value is <256, it should be
+    converted directly into decimal, 12.
+
+    https://github.com/kubernetes/kubernetes/issues/45024
+    """
+    pretreated_scsilun = int(lun, 16)
+    if pretreated_scsilun < 256:
+        return pretreated_scsilun
+    else:
+        return (pretreated_scsilun >> 16 & 0xFFFF) | \
+               (pretreated_scsilun & 0xFFFF) << 16
+
+
+class RESTClient(object):
+    """
+    driver side client. Used to interaction with pyds8k client as an adaptor.
+
+    !--important: the id field of all resources are case insensitive--!
+    """
+
+    def __init__(self, service_address, user, password,
+                 port=None,
+                 hostname='',
+                 ):
+        client_kwargs = {'service_address': service_address,
+                         'user': user,
+                         'password': password,
+                         }
+        if port:
+            client_kwargs.update({'port': port})
+        if hostname:
+            client_kwargs.update({'hostname': hostname})
+
+        self.client = Client(**client_kwargs)
```

Comment:
Is there a way to encrypt the connection (e.g. HTTPS)? it looks like plain HTTP.
With XIV we use SSL, and with SVC we use SSH...

Reply:
https

Reply:
https://github.com/IBM/pyds8k/blob/develop/pyds8k/httpclient.py

Reply:
Got it. `secure=True`